tests/extrusion/test_real_tensorproduct.py

```diff
@@ -0,0 +1,130 @@
+"""Tests for (foo x Real) exturded spaces"""
+import numpy as np
+import pytest
+
+from firedrake import *
+
+
+@pytest.mark.parametrize(('testcase', 'tolerance'),
+                         [(("CG", 1, 'lin'), 1e-15),
+                          (("CG", 1, 'sin'), 1e-5)])
+def test_vertical_average(extmesh, testcase, tolerance):
+    """Test projection on P1 x Real space"""
+    hfamily, hdegree, functype = testcase
+    vfamily, vdegree = "Real", 0
+
+    mesh = extmesh(2, 2, 8)
+    fspace = FunctionSpace(mesh, hfamily, hdegree, vfamily=vfamily, vdegree=vdegree)
+
+    u = TrialFunction(fspace)
+    v = TestFunction(fspace)
+
+    xyz = SpatialCoordinate(mesh)
+    if functype == 'sin':
+        expr = sin(pi*xyz[2])
+        solution_expr = Constant(2.0/pi)
```

Comment:
This is a bit icky. Why not just put the expression and solution in the pytest parametrisation?

Reply:
Could do that too.

Reply:
I can modify this to use string expressions if it's better.

Reply:
Use parametrised fixtures

```python
@pytest.fixture
def V(extmesh):
    mesh = extmesh(2, 2, 8)
    return FunctionSpace(mesh, "CG", 1, vfamily="Real", vdegree=1)

@pytest.fixture(params=["linear", "sin"]):
def variant(request):
    return request.param

@pytest.fixture
def expr(variant, V):
    x, y, z = SpatialCoordinate(V.ufl_domain())
    if variant == "linear":
        return z
    else:
       return sin(pi*z)

@pytest.fixture
def solution(variant):
    return {"linear": Constant(0.5), "sin": Constant(2.0/pi)}[variant]

@pytest.fixture
def tolerance(variant):
    return {"linear": 1e-15, "sin": 1e-5}[variant]

def test_foo(V, expr, solution, tolerance):
   ...
```

Reply:
Huh, okay. Will do.